backend/saves.py

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python
+from dataclasses import dataclass
+from sqlite3 import connect
+import json
+
+@dataclass
+class Saves():
+#
+    DBPath: str
+#
+    RMTABLE = 'drop table if exists games'
+    MKTABLE = 'create table games (id varchar(3) primary key, data json)'
+#
+    GET_ALL = 'SELECT id, json_extract(data, "$") as data from games'
+    GET_ID = 'SELECT id, data from games where id = ? '
+#
+    DELETE = 'DELETE FROM games WHERE id = ?'                      
+    UPDATE = 'INSERT OR REPLACE INTO games values(?, ?) '
+#
+
+    def _dictate(self, cursor, row):
+        fields = [column[0] for column in cursor.description] 
+        return {key: value for key, value in zip(fields, row)}
+
+    def _connect(self):
+        conn = connect(self.DBPath)
+        conn.row_factory = self._dictate
+        return conn
+
+    def _initdb(self):
```

Comment:
If we wanted to move this into Starlette we could do something along the lines of:
```python
# Maybe in database.py file
from sqlite3 import PARSE_DECLTYPES, connect

db = connect('casino.db', detect_types=PARSE_DECLTYPES, check_same_thread=False)
db.row_factory = lambda c, r: {k: v for k, v in zip([cl[0] for cl in c.description], r)} 
c = db.cursor()


def create_tables():
    # We can use IF NOT EXISTS to only create the table if we need to
    c.execute(
        '''
        CREATE TABLE IF NOT EXISTS game (
            id          INTEGER PRIMARY KEY AUTOINCREMENT,
            created     DATETIME DEFAULT CURRENT_TIMESTAMP,
            modified    DATETIME DEFAULT CURRENT_TIMESTAMP,
            name        TEXT NOT NULL
        );
        '''
    )
    c.execute(
        '''
        CREATE TABLE IF NOT EXISTS round (
            id          INTEGER PRIMARY KEY AUTOINCREMENT,
            created     DATETIME DEFAULT CURRENT_TIMESTAMP,
            modified    DATETIME DEFAULT CURRENT_TIMESTAMP,
            game_id     INTEGER,
            state       JSON,
             
            FOREIGN KEY (game_id) REFERENCES game(id)
        );
        '''
    )
    c.execute(CREATE INDEX IF NOT EXISTS game_idx ON round(game_id))
    # Other tables... if we need them in the future User, Points, etc...
```
    
Then in the `api.py` where we initialize the app we can take advantage of a few callbacks.
```python
app = Starlette(
    debug=True,
    routes=routes,
    on_startup=[create_tables],
    on_shutdown=[],
)
```

Reply:
thank you very much for the detailed suggestion. I applied this and hooked into a 'save' route and function to make sure I could get that to do db write.